Introduce `withSyncEvent` action wrapper utility and proxy `event` object whenever it is not used

felixarntz opened the pull request:
This implements the 1st step from #64944. Please see the issue for additional context on the rationale and overarching approach for this.

## Relevant technical choices

* In order for directive implementations to access resolved data from the entries, a few new functions are needed. This is because the `evaluate` function resolves the entry into the relevant data, but then also immediately evaluates it.
    * For the case of an event listener callback, this means the `evaluate` function will not only "look up" the callback, but also call it.
    * Since we need to conditionally alter the `event` object based on whether the callback is wrapped in `withSyncEvent` prior to calling the callback, the `evaluate` function is not sufficient.
    * In a future PR, we would then also need to call `splitTask()` before running the callback function (see 2nd step from #64944).
* As a solution, this PR introduces two new functions `resolveEntry` and `evaluateResolved`. If they are called right after each other, it's effectively the same as `evaluate`. In fact, this PR updates `evaluate` to rely on these two lower-level functions, to avoid duplicate code.
* Most directives can still use `evaluate` as is, which is more ergonomic. However the `on` (but not `on-async`) directives require the more granular access of the callback first, to only execute it later.

## Note to reviewers

Because several files are touched, but only _some_ specific changes are non-trivial, I would advise to review the commits individually at first. That should help convey the bigger picture than looking at the whole PR diff at once.

## Testing instructions

1. Add an image block to a post or page and enable the lightbox for it.
2. In the frontend, after opening the lightbox, navigate using the `Tab` key and ensure no warnings are present in the browser console.
    * This is because the `handleKeydown` action of the `core/image` block's store passes its callback through `withSyncEvent`.
3. You can now alter the code to see what would happen if the `handleKeydown` action was doing it wrong: Edit `packages/block-library/src/image/view.js` and remove the `withSyncEvent` usage from the `handleKeydown` callback, so that it is no longer wrapped. Don't forget to rerun the JS build.
4. Repeat the frontend interaction from step 2, and you should see a warning in the browser console that `withSyncEvent` needs to be used.
5. Add a search block to a post or page and set its button position to "button only".
6. In the frontend, click the Search button and ensure no warnings are present in the browser console.
    * This is because the `openSearchInput` action of the `core/search` block's store passes its callback through `withSyncEvent`.
7. Similar to step 3, edit `packages/block-library/src/search/view.js` and remove the `withSyncEvent` usage from the `openSearchInput` callback, so that it is no longer wrapped. Don't forget to rerun the JS build.
8. Repeat the frontend interaction from step 6, and you should see a warning in the browser console that `withSyncEvent` needs to be used.

Commit:
Implement withSyncEvent action wrapper utility.

Commit:
Prepare Interactivity API infrastructure for awareness of action prior to evaluating it.

gziolo commented:
> So far, I have not implemented the actual proxying yet. I would like to get a sense of whether this minor expansion of the Interactivity API infrastructure makes sense before continuing.

I don't know the codebase, but you are on track as much as I understand the changes and their implications. I also wanted to share that it might not be easy to get more feedback in the next 2-3 weeks as many folks, myself included, take longer holiday breaks 🎉

michalczaplinski reviewed (commented):
I gave it an initial review and so far it's looking good. The approach of splitting `evaluate` into 2 functions seems sound to me. I see that you haven't implemented the proxying yet, so I'm happy to take another look once that's closer to being ready.

PS. The Interactivity API is almost entirely tested using the e2e suite in `test/e2e/specs/interactivity/`. For this PR, you'll probably need to modify some of the existing "test" blocks like [`packages/e2e-tests/plugins/interactive-blocks/directive-on/view.js`](https://github.com/r-chrzan/gutenberg/blob/6490514a4e5fd14635de1faaa6d395201e248620/packages/e2e-tests/plugins/interactive-blocks/directive-on/view.js) or create a new "test" block in that folder.

Commit:
Merge branch 'trunk' into add/64944-with-sync-event

Commit:
Proxy event object when withSyncEvent() is not used.

felixarntz commented:
I have implemented the event proxying logic. For now, I'm warning about accessing the following without `withSyncEvent()`:
* `event.currentTarget`
* `event.preventDefault()`
* `event.stopImmediatePropagation()`
* `event.stopPropagation()`

As discussed on the issue, there may be more properties and methods to warn about, but we could always expand later. I think these 4 are a good starting point and should cover what's most common.

Review much appreciated!

felixarntz commented:
Any chance I can get 1-2 reviews for the PR this week? That would be super helpful so we could ideally land this in Gutenberg soon. Thank you! 🙌

Commit:
Merge branch 'trunk' into add/64944-with-sync-event

Commit:
Ensure generator functions using withSyncEvent() are wrapped correctly to still be recognized as generator functions.

Commit:
Update Interactivity API documentation to reference withSyncEvent().

Commit:
Use withSyncEvent() in all built-in actions that require it.

felixarntz commented:
Some updates:
* In https://github.com/WordPress/gutenberg/pull/68097/commits/21e51be1298a268b175a8f1d7d4daf6af6ab1361 I fixed support for generator functions so that they're still recognized as such even when wrapped.
* In https://github.com/WordPress/gutenberg/pull/68097/commits/a582f01b2308dc58aa66e072f712cdfa3379f88f I updated the Interactivity API docs accordingly to include `withSyncEvent()`.
* In https://github.com/WordPress/gutenberg/pull/68097/commits/5d68fcb8226bb2dbc7473757bd5c586df3ee3654 I updated all relevant built-in actions to use `withSyncEvent()` as necessary.
    * This of course adds a bunch more files changed to the PR, so I'd recommend reviewing the commits individually at this point, as it makes things easier.
    * If you prefer that we do this separately, we can also temporarily revert this commit.

Commit:
Merge branch 'trunk' into add/64944-with-sync-event

Commit:
Minor fixes for withSyncEvent docs.

felixarntz commented on docs/reference-guides/interactivity-api/api-reference.md:
```diff
@@ -1253,6 +1256,43 @@ store( 'mySliderPlugin', {
 
+### withSyncEvent()
+
+Actions that require synchronous access to the `event` object need to use the `withSyncEvent()` function to wrap their handler callback. This is necessary due to an ongoing effort to handle store actions asynchronously by default, unless they require synchronous event access. Therefore, as of Gutenberg `TODO: Add release number here!` / WordPress 6.8 all actions that require synchronous event acc ...
```
Note: The `TODO` here can only be addressed once this is ready for merge and we know which Gutenberg version this will be released in.

westonruter commented on docs/reference-guides/interactivity-api/api-reference.md:
````diff
@@ -885,16 +887,17 @@ function splitTask() {
 ```
 
-You may want to add multiple such `yield` points in your action if it is doing a lot of work.
+You may notice the use of the [`withSyncEvent()`](#withsyncevent) utility function in this example. This is necessary due to an ongoing effort to handle store actions asynchronously by default, unless they require synchronous event access. Otherwise a deprecation warning will be triggered, and in a future release the behavior will change accordingly.
````
```suggestion
You may notice the use of the [`withSyncEvent()`](#withsyncevent) utility function in this example. This is necessary due to an ongoing effort to handle store actions asynchronously by default, unless they require synchronous event access (which this example does due to the call to `event.preventDefault()`). Otherwise a deprecation warning will be triggered, and in a future release the behavior will change accordingly.
```

westonruter commented on packages/interactivity/src/utils.ts:
```diff
@@ -374,3 +374,23 @@ export const isPlainObject = (
+ * @param callback The event callback.
+ * @return Wrapped event callback.
+ */
+export const withSyncEvent = ( callback: Function ): Function => {
```
I think it's preferable to use a function rather than an anonymous function so that the function name will more reliably appear in the call stack.

```suggestion
export function withSyncEvent( callback: Function ): Function => {
```

That said, it seems (in [consulting with Gemini](https://g.co/gemini/share/a1954db83a90)) modern debuggers are able to guess the name of the anonymous function based on context. Still, something explicit seems better.

felixarntz replied:
I was looking at the functions above which were using `const`, but fair point. Looks like most of the file is using functions and I agree that's better. Updating...

felixarntz replied:
Done as part of https://github.com/WordPress/gutenberg/pull/68097/commits/bf3dbcc861367a38131ac0943ad49f60e2474912

westonruter commented on packages/interactivity/src/utils.ts:
```diff
@@ -374,3 +374,23 @@ export const isPlainObject = (
+ * @return Wrapped event callback.
+ */
+export const withSyncEvent = ( callback: Function ): Function => {
+	if ( callback?.constructor?.name === 'GeneratorFunction' ) {
+		const wrapped = function* ( ...args: any[] ) {
+			yield* callback( ...args );
+		};
+		wrapped.sync = true;
+		return wrapped;
+	}
+
+	const wrapped = ( ...args: any[] ) => callback( ...args );
+	wrapped.sync = true;
+	return wrapped;
```
Removing a bit of code duplication:

```suggestion
	let wrapped;
	if ( callback?.constructor?.name === 'GeneratorFunction' ) {
		wrapped = function* ( ...args: any[] ) {
			yield* callback( ...args );
		};
	} else {
		wrapped = ( ...args: any[] ) => callback( ...args );
	}
	wrapped.sync = true;
	return wrapped;
```

felixarntz replied:
I had tried to do this initially in the same way, but it doesn't work:
* The `let wrapped;` will require a type.
* Setting it to `Function` will cause the `.sync` assignment to not work because that's not available on a function object.
* Setting it to `any` would work, but not be great because `any`.

I just searched a bit further and found https://stackoverflow.com/questions/12766528/build-a-function-object-with-properties-in-typescript. That seems to work better, so I'm gonna go with that.

felixarntz replied:
Done as part of https://github.com/WordPress/gutenberg/pull/68097/commits/bf3dbcc861367a38131ac0943ad49f60e2474912.

I also identified a bug where the additional `sync` property was not being passed through, due to passing all store actions through the `withScope()` function, which happens via the `proxifyStore()` function.

By introducing an extra type `SyncAwareFunction`, this can be handled a bit more elegantly, as it allows to simply cast regular functions into that type and then set the `sync` property as needed (in practice, it's always `true`, or it's not set, i.e. a regular function).

I have tested this approach and it works well. It also ensures the original underlying JS types (e.g. `Function` or `GeneratorFunction`) are maintained.

westonruter commented on packages/interactivity/src/directives.tsx:
```diff
@@ -50,6 +50,39 @@ function deepClone< T >( source: T ): T {
 	return source;
 }
 
+function wrapEventAsync( event: Event ) {
```
Add jsdoc for this?

felixarntz replied:
Done in https://github.com/WordPress/gutenberg/pull/68097/commits/2dd6eabe1a7d7ade2f3ff1834dd7232551498e72

westonruter commented on packages/interactivity/src/directives.tsx:
```diff
@@ -50,6 +50,39 @@ function deepClone< T >( source: T ): T {
 
+function wrapEventAsync( event: Event ) {
+	const handler = {
+		get( target: any, prop: any, receiver: any ) {
```
```suggestion
		get( target: Event, prop: string, receiver: any ) {
```

Maybe the typing of `receiver` could be made more specific as well.

felixarntz replied:
Done in https://github.com/WordPress/gutenberg/pull/68097/commits/682ee6f3fb8647e664287df73c77ce79d6cc3e1a

I used `string | symbol` for `prop`, in accordance with other proxy handlers already present in Interactivity API code. Those handlers also use `receiver: any`, so I think it's fine to leave that as-is.

westonruter commented on packages/interactivity/src/directives.tsx:
```diff
@@ -50,6 +50,39 @@ function deepClone< T >( source: T ): T {
 	return source;
 }
 
+function wrapEventAsync( event: Event ) {
```
```suggestion
function wrapEventAsync( event: Event ): Proxy<Event> {
```

felixarntz replied:
This doesn't appear to work: `'Proxy' refers to a value, but is being used as a type here.`

westonruter commented on packages/interactivity/src/utils.ts:
```diff
@@ -374,3 +374,23 @@ export const isPlainObject = (
+		return wrapped;
+	}
+
+	const wrapped = ( ...args: any[] ) => callback( ...args );
```
Should this be preserving `this` context like you're doing in `wrapEventAsync`?

```suggestion
	const wrapped = function ( ...args: any[] ) {
		return callback.apply( this, args )
	};
```

westonruter replied:
Same for the generator variant above.

felixarntz replied:
Done as part of https://github.com/WordPress/gutenberg/commit/bf3dbcc861367a38131ac0943ad49f60e2474912

westonruter commented on packages/interactivity/src/directives.tsx:
```diff
@@ -50,6 +50,39 @@ function deepClone< T >( source: T ): T {
+			if ( value instanceof Function ) {
+				return function ( this: any, ...args: any[] ) {
+					return value.apply(
+						this === receiver ? target : this,
```
What does `this === receiver` mean here? Why wouldn't it only use `this`?

felixarntz replied:
See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy#no_private_property_forwarding

Commit:
Merge branch 'trunk' into add/64944-with-sync-event

felixarntz committed:
Clarify documentation.

Co-authored-by: Weston Ruter <westonruter@google.com>

Commit:
Merge branch 'add/64944-with-sync-event' of github.com:WordPress/gutenberg into add/64944-with-sync-event

Commit:
Enhance withSyncEvent implementation and ensure the sync flag is maintained when proxying functions via withScope.

Commit:
Add doc block for wrapEventAsync().

Commit:
Use more specific types for event proxy handler.

felixarntz commented:
I have done some extensive testing on the code changes and added testing instructions to the PR description. I identified one crucial bug, which was fixed (see https://github.com/WordPress/gutenberg/pull/68097#discussion_r1934812028). But everything is working well now, and you can work through the testing instructions to validate.

westonruter reviewed (approved):
LGTM, but I defer to deeper review from the Interactivity API team.

felixarntz commented:
Any chance you can take another look at this @michalczaplinski?

luisherranz commented on packages/interactivity/src/utils.ts:
```diff
@@ -374,3 +392,26 @@ export const isPlainObject = (
+
+	wrapped.sync = true;
+	return wrapped;
+}
```
Why don't you simply add the `sync` property to the original function? Do you want to prevent modifying it?

```suggestion
export function withSyncEvent( callback: Function ): SyncAwareFunction {
	( callback as SyncAwareFunction ).sync = true;
	return callback as SyncAwareFunction;
}
```

felixarntz replied:
Great idea! I thought initially that wrapping it was safer than directly altering it, but since the `callback` function would be usually just declared anonymously and immediately passed to `withSyncEvent`, there's probably no side effects of directly altering it.

Most importantly, I agree with your suggestion because it makes the logic much simpler and more error-proof, maintaining whichever type the original function had before passing it to `withSyncEvent`, instead of trying to recreate/guess it.

I've amended this in https://github.com/WordPress/gutenberg/pull/68097/commits/f0fa92afaaca9b625a341610564c73f0fb9ad2a3, alongside minor wording adjustments based on the change.

westonruter replied:
> Great idea! I thought initially that wrapping it was safer than directly altering it, but since the `callback` function would be usually just declared anonymously and immediately passed to `withSyncEvent`, there's probably no side effects of directly altering it.

The key word here is _usually_, right? You could define a named function in the module and then reference it here in `withSyncEvent()`. What if you then use that same function for an async event? Then it would eventually result in that code not being able run asynchronously, since the function has the `sync` flag set on it.

felixarntz replied:
Yeah I agree it wouldn't be 100% guaranteed. But given the wrapping requires other quirks that also don't make it 100% reliable (not every function is a plain function or has constructor `GeneratorFunction`), I think avoiding the wrapping is safer for the vast majority of use-cases.

Specifically to your example, I have a hard time seeing that being common: First, because if the function is the same and needs synchronous `event` access, why would it elsewhere _not_ access a synchronous `event` property? Second, all the Interactivity API examples and Core code define the actions right on the object, so it's not encouraged to do it differently anywhere.

luisherranz replied:
I also think that the likelihood of the same function being executed synchronously and asynchronously at the same time must be really low, because it would have to include conditional code and depend on the place of the DOM where it is executed. It is not impossible and I'm sure there is some case, but in that case, maybe it would even be safer to always run it synchronously.

That said, I don't oppose leaving that part of the code as it is. Up to you 🙂

luisherranz reviewed (changes requested):
Looks good to me! 🙂 

I only have one request: instead of creating two separate functions, we can pass an options object to `evaluate`. It's a pattern we're already using in other parts of the Interactivity API. This way, we not only simplify the code, but we also open the door to adding more options in the future in a simple way.

As `evaluate` is not publicly exposed yet, we can introduce breaking changes without any issues. 

Perhaps we could start with these two options:

```js
const result = evaluate( entry, { 
  resolveFunctions: false,
  arguments: [ ... ], // Arguments passed to the resolved function.
} );
```

Or we could even not add the option of arguments yet and, if someone needs to pass arguments, use the option `resolveFunctions: false`:

```js
const cb = evaluate( entry, { resolveFunctions: false } );
cb( event );
```

EDIT: I guess e2e tests can be added once we move from deprecation to error, right? What do you think?

Commit:
Merge branch 'trunk' into add/64944-with-sync-event

Commit:
Amend callback in withSyncEvent instead of wrapping it.

felixarntz commented:
@luisherranz Thanks for the review!

> I only have one request: instead of creating two separate functions, we can pass an options object to `evaluate`. [...]

Hmm, I'm not sure about that idea, given that the `resolveFunctions` option would significantly change what the function does and give it different code paths. Boolean arguments like this tend to increase complexity, see for instance https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3 which goes in depth about this. What would be the problem of having the two lower-level functions that `evaluate` is then composed of?

> I guess e2e tests can be added once we move from deprecation to error, right? What do you think?

For what kind of interactions would you like to see e2e tests? Not arguing against it at all, just wondering since I'm personally not very familiar with writing e2e tests.

luisherranz commented:
> given that the resolveFunctions option would significantly change what the function does and give it different code paths. Boolean arguments like this tend to increase complexity
> [...]
> What would be the problem of having the two lower-level functions that evaluate is then composed of (especially given none of them are exposed)?

For me, it's fine if we increase the complexity in the runtime of the Interactivity API to simplify the external APIs that the extenders will use. Keep in mind that, although it is still far off in the roadmap, the idea is to let the extenders create their own directives someday. So for them, in my opinion, it will be more complex to use multiple functions than to pass options to a single function, especially if we keep adding more options in the future.

luisherranz commented:
> For what kind of interactions would you like to see e2e tests? Not arguing against it at all, just wondering since I'm personally not very familiar with writing e2e tests.

Basically, a test that verifies that if one of the synchronous functions is executed without the wrapper, the deprecation is shown correctly. We can leave it for the next phase, anyway.

felixarntz commented:
@luisherranz 
> For me, it's fine if we increase the complexity in the runtime of the Interactivity API to simplify the external APIs that the extenders will use. Keep in mind that, although it is still far off in the roadmap, the idea is to let the extenders create their own directives someday. So for them, in my opinion, it will be more complex to use multiple functions than to pass options to a single function, especially if we keep adding more options in the future.

I think at this point this is mostly a theoretical concern. At the moment, none of these functions is externally available, so introducing a boolean flag would only increase complexity. So for that reason I don't think it makes sense to change it now, and we could revisit this again if and once we make those functions externally available. The way it currently works in this PR could easily be adjusted later if we decide to.

> Basically, a test that verifies that if one of the synchronous functions is executed without the wrapper, the deprecation is shown correctly. We can leave it for the next phase, anyway.

👍  Makes sense, thanks!

felixarntz commented:
> I don't like the idea of leaving things in a state that will require changes later. The fewer adjustments we have to make once this is public, the better.

Fair point, let's try to figure it out now. You mentioned before:

> So for them, in my opinion, it will be more complex to use multiple functions than to pass options to a single function, especially if we keep adding more options in the future.

Going back to my previous point: Using boolean flags increases the complexity of the code. This also impacts those that use the function(s). Today, if you see the `evaluate` function, you know what it does. If we make its behavior conditional on a boolean flag, you'll have to look more closely, since `evaluate` no longer serves one clear purpose. So IMO having separate functions for separate purposes helps also users of the eventually external function(s).

But let's take a step back first, I have a follow up questions to your idea from https://github.com/WordPress/gutenberg/pull/68097#pullrequestreview-2605237411: How would `resolveFunctions` change what `evaluate` does? I don't fully understand how you intend it to be used, probably just a terminology thing to clarify. In this PR I implemented `resolveEntry` and `evaluateResolved`, both of which (in that order) make up the existing `evaluate` function. If I call `evaluate( { resolveFunctions: false } )`, which of the above two inner functions should be executed?

luisherranz commented:
> If I call `evaluate( { resolveFunctions: false } )`, which of the above two inner functions should be executed?

What do you mean? This is trivial to implement on top of the original function: https://github.com/WordPress/gutenberg/commit/3dc8989579e86e1b9eccfa2b983ad9d4f9a1cc9f

Commit:
Merge branch 'trunk' into add/64944-with-sync-event

felixarntz commented:
> > If I call `evaluate( { resolveFunctions: false } )`, which of the above two inner functions should be executed?
> 
> What do you mean? This is trivial to implement on top of the original function: [3dc8989](https://github.com/WordPress/gutenberg/commit/3dc8989579e86e1b9eccfa2b983ad9d4f9a1cc9f)

Your code clarifies what I was asking before: Before you were using the name `resolveFunctions`, so that's what was unclear in what you meant for it to do. The name you used now `invokeFunctions` makes more sense to me, so now I see what you mean. 👍 

Regarding https://github.com/WordPress/gutenberg/commit/3dc8989579e86e1b9eccfa2b983ad9d4f9a1cc9f, I think this could work in principle, but I see two problems with it that the current PR implementation and `trunk` don't have:
* You're getting rid of `...args` which is currently being passed to the function. That would cause problems because e.g. `event` would no longer be able to be passed through to the callback, which would still need to happen if `options.invokeFunctions` is `true`.
* In case the resolved entry is a function, your code ignores the negation operator that may be present. This is problematic because then the caller of `evaluate( entry, { invokeFunctions: false } )` has no idea whether after calling the returned function themselves is needs to be negated or not.

To be clear, I'm not fundamentally opposed to your suggestion, but these two drawbacks would need to be addressed to make it beneficial over the current implementation.

luisherranz commented:
> so that's what was unclear in what you meant for it to do. The name you used now `invokeFunctions` makes more sense to me, so now I see what you mean. 👍

Sorry! My bad. I'm glad it's clear now 🙂

> You're getting rid of `...args` which is currently being passed to the function. That would cause problems because e.g. `event` would no longer be able to be passed through to the callback, which would still need to happen if `options.invokeFunctions` is `true`.

See [this comment](https://github.com/WordPress/gutenberg/pull/68097#pullrequestreview-2605237411).

But... I've been thinking, and I think we should make things right and stop invoking functions in `evaluate`. It's not only not needed, but actually an antipattern. We can take this opportunity to make things right and actually educate people about how to correctly use the Interactivity API.

A bit of context: I should have made `evaluate` not invoke functions by default when we made the change to the [new Store API](https://github.com/WordPress/gutenberg/discussions/53586). It's basically a leftover, because before the new Store API, the derived state was done with functions (selectors) instead of getters:

```js
// Old Store API
store({
  myPlugin: {
    state: {
      someNumber: 1,
    },
    selectors: {
      sumNumbers: ({ state, context }) => {
        return state.someNumber + context.someNumber;
      },
    },
    actions: {
      someAction: (store) => {
        const { selectors } = store;
        selectors.sumNumbers(store);
      },
    },
  },
});

// New Store API
const { state } = store("myPlugin", {
  state: {
    someNumber: 1,
    get sumNumbers(): number {
      const { someNumber } = getContext();
      return state.someNumber + someNumber;
    },
  },
  actions: {
    someAction: () => {
      state.sumNumbers;
  },
});
```

```html
// With old Store API
<span data-wp-text="myPlugin.selectors.sumNumbers"></span>

// With new Store API
<span data-wp-text="state.sumNumbers"></span>
```

With the new Store API, there's no need for all the directives that need a value, like `data-wp-text`, `data-wp-class`, `data-wp-bind`, etc, to invoke functions. Actually, it means people are doing it wrong, because they are still using some sort of _selectors_ instead of proper derived state with getters. The only directives that need to invoke functions are the ones that are supposed to receive a callback, like `data-wp-bind`, `data-wp-init`, `data-wp-watch`, etc, and they can easily do so by invoking the callback themselves:

```js
const cb = evaluate( entry );
cb( ...args ); // Pass whatever args they need.
```

The problem is that, from time to time, people who still don't know how to use the Interactivity API well yet, use functions instead of getters, and as they work fine, they keep using them ([for example](https://github.com/WordPress/gutenberg/discussions/55642#discussioncomment-9933614)).

So, to make this 100% right, I'm thinking now that we should follow the same strategy used here for async actions and actually deprecate the fact that directives that are meant to receive a value invoke functions at all. 

Of course, that is out of scope for this PR, but maybe we can take the first steps here? It should be something as simple as this: https://github.com/WordPress/gutenberg/commit/f0525ed790d96048c6188f2f64ff5804f6bb7ca4

Let me know what you think 🙂

Commit:
Merge branch 'trunk' into add/64944-with-sync-event

felixarntz commented:
> But... I've been thinking, and I think we should make things right and stop invoking functions in `evaluate`. It's not only not needed, but actually an antipattern. We can take this opportunity to make things right and actually educate people about how to correctly use the Interactivity API.
> 
> A bit of context: I should have made `evaluate` not invoke functions by default when we made the change to the [new Store API](https://github.com/WordPress/gutenberg/discussions/53586). It's basically a leftover, because before the new Store API, the derived state was done with functions (selectors) instead of getters:
> 
> [...]
> 
> So, to make this 100% right, I'm thinking now that we should follow the same strategy used here for async actions and actually deprecate the fact that directives that are meant to receive a value invoke functions at all.

That makes a lot of sense now. Thanks for explaining in more detail!

> Of course, that is out of scope for this PR, but maybe we can take the first steps here? It should be something as simple as this: [f0525ed](https://github.com/WordPress/gutenberg/commit/f0525ed790d96048c6188f2f64ff5804f6bb7ca4) (not tested)

This looks good to me for the most part. The only thing I find a bit confusing for `evaluate()` is that the negation operator is only used if it's not a function. It makes sense based on what you're saying, that functions should only be used for actions and callbacks, so they would never have a return value and therefore it would be irrelevant to use a negation operator on a function. But today, someone might still do something like `data-wp-show="! actions.getHiddenState()"`. Obviously that's bad for many reasons, and I don't anticipate for this to be common. But how should we deal with it?

I think the most straightforward solution _for this PR_ is to continue invoking functions in `evaluate()` _only if there is also a negation operator present_. In this case, we can be sure that this is used in a directive other than `data-wp-on...`, `data-wp-watch...` etc., since those would never expect a return value. And when that happens, we should trigger a deprecation warning, to indicate that this will no longer work in the near future.

In other words there would be three possible scenarios for an entry for `evaluate()`:
* Entry is not for a function --> **handle like today**
* Entry is for a function with negation operator --> **handle like today, but trigger deprecation warnings**
* Entry is for a function without negation operator --> **return function without calling it**

By doing that, we don't (immediately) break the pattern of `! function()` that someone may be using, but we warn about it, and we would have a plan to remove support for it.

Then we can have a separate PR for a later Gutenberg release for WordPress 6.9 that removes `...args` from the `evaluate()` function entirely and removes support for the negation operator if it is provided in combination with a function.

Commit:
Revert "Prepare Interactivity API infrastructure for awareness of action prior to evaluating it."

This reverts commit dba93ec4f7bb617d1f12ba1a06acfa39ff33d4b3.

Commit:
Update evaluate() to no longer invoke functions (except where needed for BC) and move responsibility to the caller.

felixarntz commented:
@luisherranz I reverted the addition of the two new lower-level functions and instead implemented in https://github.com/WordPress/gutenberg/pull/68097/commits/0dbd15f7abc1a214ad4bfad677a32e204a2bdd77 the change of no longer invoking functions in `evaluate()`, except for backward compatibility in case a negation operator is present (which will now trigger a deprecation warning).

This is similar to the example code you provided in https://github.com/WordPress/gutenberg/commit/f0525ed790d96048c6188f2f64ff5804f6bb7ca4, just with a few additions to prevent errors, e.g. we have to check for whether the `evaluate()` return value is a function before calling it.

Having these checks everywhere is a bit annoying in terms of our codebase, so ideally we can improve this in the future. One of the reasons for those checks is that `evaluate()` is generic and returns `any`, so maybe we would want to introduce two wrappers for it, one that guarantees to return a function and another that guarantees to return anything _but_ a function. Then the correct variant could be used in each directive as appropriate. It might also make the API easier to use externally. But as you said before, for now this would be mostly about cleaning up our own code, so we can probably do that in a separate PR.

The overall behavior remains the same as before, it's just a different implementation under the hood.

luisherranz commented:
> someone might still do something like `data-wp-show="! actions.getHiddenState()"`. Obviously that's bad for many reasons, and I don't anticipate for this to be common. But how should we deal with it?

`! actions.getHiddenState()` is invalid as an iAPI reference, because it points to `store[' actions']['getHiddenState()']`.

If what you meant was `!actions.getHiddenState`, the way we should treat it is with a deprecation warning indicating that state should be used instead (either global, local or derived): so `!context.hidden` or `!state.hidden`.


> In other words there would be three possible scenarios for an entry for `evaluate()`:
> 
> - Entry is not for a function --> **handle like today**
> - Entry is for a function with negation operator --> **handle like today, but trigger deprecation warnings**
> - Entry is for a function without negation operator --> **return function without calling it**

I understand this logic well.

luisherranz committed:
Export withSyncEvent

luisherranz commented:
> Having these checks everywhere is a bit annoying in terms of our codebase, so ideally we can improve this in the future

Indeed. Right now, there isn't much consistency in how errors are being captured and reported. It's something we need to look at carefully and if possible, delegate to the Interactivity API itself (instead of the directives).

> It looks like something in this change is causing all sorts of e2e tests for the Interactivity API to fail. Maybe you find something in my implementation that's off and can easily be fixed.

It seems you had simply forgotten to export `withSyncEvent`. I'm going to push to see if the tests pass with that change.

Commit:
Merge branch 'trunk' into add/64944-with-sync-event

felixarntz commented:
> > someone might still do something like `data-wp-show="! actions.getHiddenState()"`. Obviously that's bad for many reasons, and I don't anticipate for this to be common. But how should we deal with it?
> 
> `! actions.getHiddenState()` is invalid as an iAPI reference, because it points to `store[' actions']['getHiddenState()']`.
> 
> If what you meant was `!actions.getHiddenState`, the way we should treat it is with a deprecation warning indicating that state should be used instead (either global, local or derived): so `!context.hidden` or `!state.hidden`.

Yes, that's what I meant. I agree with what you're saying. But I don't think this PR is the right place to do that because it is a separate issue to address. The change here is already quite substantial, so broader changes like this would be better handled in a separate PR to separately review.

For this PR specifically, I addressed your feedback to not introduce two new functions, but instead alter `evaluate` to not invoke functions (except for the one backward compatibility consideration), and I think that sets us up in the right direction.

> EDIT: Oh, I think I understand it better now: the difference is that you would trigger the deprecation warnings in the `evaluate` function, but I was thinking of putting them inside the directives that shouldn't use functions but derived state (as in my [example here](https://github.com/WordPress/gutenberg/commit/f0525ed790d96048c6188f2f64ff5804f6bb7ca4)). Is that right?
> 
> The point is that I would like to show deprecation warnings even when a function is used without a negation operator in the directives that accept a value (`class`, `bind`, `style`, `text`, `each`…):
> 
> * This is wrong: `data-wp-text="!actions/callbacks.getSomeValue"` and it should be `data-wp-text="!state/context.someValue"`.
> * But this is also wrong: `data-wp-text="actions/callbacks.getSomeValue"` and it should be `data-wp-text="state/context.someValue"`.

I agree on all points. But see my above feedback, implementing this is unrelated to the goal of this PR. So I think it should be implemented separately. I'm happy to follow up on this and implement it myself. But I don't think it would be right to make those changes as part of this PR.

Commit:
Fix evaluate to return scoped function and always reset scope.

Commit:
Update custom directives for e2e tests to account for evaluate behavior change.

felixarntz commented:
> It seems you had simply forgotten to export `withSyncEvent`. I'm going to push to see if the tests pass with that change.

🤦 That must have gotten lost during the revert of the older commit. Thanks for catching that!

There were still several other e2e test failures after the fix (see https://github.com/WordPress/gutenberg/actions/runs/13395813918/job/37415759288?pr=68097), but I was able to identify why these were happening:
1. We can't simply return the function from `evaluate` as is because then it wouldn't be invoked with the correct scope. That's why in https://github.com/WordPress/gutenberg/pull/68097/commits/d2a990a993d6563940fe3eb5e2b9a0f85efbaa39 I fixed this to return the wrapped function in a way that it will always run with the correct scope. Part of the problem was also that the `resetScope()` call was missing in a few places before the early returns.
2. The other bit was simply to adjust several custom directives for e2e tests based on the `evaluate` behavioral change, which I fixed in https://github.com/WordPress/gutenberg/pull/68097/commits/a02b87378e98045aaefbc44e3799a65283e32dd0. Again, I don't like how this is currently handled, but since this is all internal code and the functions aren't exposed externally, we can still figure out how to make this more ergonomic in the near future I think.

These changes address all remaining e2e test failures.

luisherranz reviewed (approved):
I agree to do those deprecations in a separate PR so, I think this PR is ready now. 

Thank you very much for considering my suggestions, Felix 🙂

Commit:
Merge branch 'trunk' into add/64944-with-sync-event

Commit:
Update release version number in documentation.

felixarntz commented:
Thanks @luisherranz! I opened #69269 for us to continue discussing the best path forward for the `evaluate()` function and its usage by directives.
